docs/src/examples/numpy.rst

```diff
@@ -0,0 +1,94 @@
+.. _numpy:
+
+Conversion to NumPy and Other Frameworks
+========================================
+
+MLX array implements the `Python Buffer Protocol <https://docs.python.org/3/c-api/buffer.html>`_.
+Let's convert an array to NumPy and back.
+
+.. code-block:: python
+
+  import mlx.core as mx
+  import numpy as np
+
+  a = mx.arange(3)
+  b = np.array(a)
+  c = mx.array(b)
+
+.. note::
+
+    Since NumPy does not support ``bfloat16`` arrays, you will need to convert to ``float16`` or ``float32`` first:
+    ``np.array(a.astype(mx.float32))``.
+    Otherwise, you will receive an error like: ``Item size 2 for PEP 3118 buffer format string does not match the dtype V item size 0.``
+
+By default, NumPy copies data to a new array. This can be prevented by creating an array view:
+
+.. code-block:: python
+
+  a = mx.arange(3)
+  a_view = np.array(a, copy=False)
+  print(a_view.flags.owndata) # False
+  a_view[0] = 1
+  print(a[0].item()) # 1
+
+A NumPy array view is a normal NumPy array, except that it does not own its memory.
+This means writing to the view is reflected in the original array.
+
+While this is quite powerful to prevent copying arrays, it should be noted that external changes to the memory of arrays cannot be reflected in gradients.
+
+Let's demonstrate this in an example:
+
+.. code-block:: python
+
+  x = mx.zeros(1)
+
+  def f(x):
+      x[:] += 1
+      return x @ x
+
+  y, df = mx.value_and_grad(f)(x)
+  print("f(x) = x² =", y.item()) # 1.0
+  print("f'(x) = 2x =", df.item()) # 2.0
+
+  def g(x):
+      x_view = np.array(x, copy=False)
+      x_view[:] += 1 # modify memory without telling mx
+      return mx.array(x @ x)
```

Comment:
I would remove the pure mlx example above and change this example to

```python
x_view[:] *= x_view
return x.sum()
```

with `x` being 2s or sth to clearly show that even though the output is `x**2` the gradient will be 1s.

Reply:
Good idea. I was thinking of this example in the beginning but did not know how to express it properly.